### ISSUE TITLE
Incremental snapshots: Memory snapshot is inconsistent.

### ISSUE THREAD

**Issue by sandreim** [firecracker-microvm/firecracker#1997]
Steps to reproduce (1 snapshot):
- Start fresh microvm
- Pause microvm
- Create diff snapshot
- Kill microvm
- Start fresh Firecracker
- Load previous snapshot
- Resume guest.

Guest crashes on resume:
```
[    1.979938] BUG: stack guard page was hit at ffffc90000dfbff8 (stack is ffffc90000dfc000..ffffc90000dfffff)
[    1.980796] kernel stack overflow (double-fault): 0000 [#1] SMP PTI
[    1.981294] PANIC: double fault, error_code: 0x0
[    1.981665] CPU: 0 PID: 770 Comm: (agetty) Not tainted 4.14.55-84.37.amzn2.x86_64 #1
[    1.982324] task: ffff8801298c0e40 task.stack: ffffc90000dfc000
[    1.982351] RIP: 0010:+0x4/0x30
[    1.982352] RSP: 0018:fffffe0000007000 EFLAGS: 00010093
[    1.982353] RAX: 0000000000000000 RBX: 0000000000000001 RCX: ffffffff81600a27
[    1.982353] RDX: 0000000000000002 RSI: 0000000000000002 RDI: fffffe0000007028
[    1.982354] RBP: fffffe0000007000 R08: 0000000000000000 R09: 0000000000000000
[    1.982359] R10: 0000000000000000 R11: 0000000000000000 R12: 0000000000000000
[    1.982360] R13: 0000000000000000 R14: 0000000000000000 R15: 0000000000000000
[    1.982360] FS:  00007f7b7dbc4e00(0000) GS:ffff88012fc00000(0000) knlGS:0000000000000000
[    1.982361] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[    1.982361] CR2: fffffe0000006ff8 CR3: 0000000129a84006 CR4: 00000000003606b0
[    1.982369] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[    1.982369] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
[    1.982370] Call Trace:
[    1.982372] Code: f2 ff ff e9 f2 fc ff ff 48 8d 4d cc 41 89 c0 4c 89 e2 48 89 de 4c 89 ef e8 1a fa ff ff e9 d8 fc ff ff e8 a0 eb 00 00 55 48 89 e5 <41> 54 49 89 fc 53 48 89 f3 0f 20 d0 0f 1f 40 00 48 89 de 4c 89
[    1.982401] Kernel panic - not syncing: Machine halted.
[    1.982439] Kernel Offset: disabled
[    1.990742] Rebooting in 1 seconds..
```

Steps to reproduce (2 snapshot: base + 1 diff):
- Start fresh Firecracker
- Load a full snapshot
- Resume
- Run some commands in guest
- Pause microvm
- Create diff snapshot
- Kill microvm
- Start fresh Firecracker
- Merge base + diff
- Load merged snapshot
- Resume guest.
- Try to run fio in guest

Fio crashes:
```
root@7e47bb8f2f0a:~# fio fio4k.cfg
rand-read-4k-1: (g=0): rw=randread, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=libaio, iodepth=32
fio-3.1
Segmentation fault
```

dmesg: `[   89.934644] fio[993]: segfault at 0 ip 0000557f655bd670 sp 00007fff4b31bd48 error 6 in fio[557f6559f000+b8000]`

**Comment by aghecenco** [comment]
I have a hunch this is because the pages dirtied by the virtio device (`fio` must mean the microvm has at least a block device) are missing from the snapshot, because they're missing in turn from the KVM dirty log

**Comment by acatangiu** [comment]
To make this work we need to add support for dirty page tracking to `vm-memory`. Guest pages dirtied by the `firecracker` process (during device emulation for example) are not guaranteed to be caught/reported by KVM's dirty page tracking.

**Comment by sandreim** [comment]
Implemented in #2331.
